FLAMEGPU/FLAMEGPU2_dev
# In-Layer Concurrency

Implement and detect in-layer concurrency via streams. 

This is now ready for review. It is not compelte, but is most of the way there and provides the majority of the intended benefit, although things can be improved.

Need to check that it has worked as intended on Windows (i.e. do the tests pass) in Release mode.

## Issues

Closes #189

Outstanding improvements detailed in #449


## Completed bits 

+ [x] add `CUDAConfig::inLayerConcurrency` property to prevent stream-based concurrency
    + required for testing, and potentially useful for debugging.
+ [x] Programatic Step Timing, init, exit and RTC timing.
    + [x] Plus output via `-t/--timing`?
+ [x] Simulate() NVTX Range
+ [x] NVTX Improvements `CUDASimulation`
+ [x] Refactor Step() into smaller chunks.
    + [x] Make Init functions user-callable
    + [x] Make Exit functions user-callable
+ [x] Better CUDASteam management/re-use
+ [x] Concurrent post-processing via streams (PBM construction, Scattering, CUB/Thrust calls...)
+ Add tests which detect concurrency, through the average runtime of steps:
    + [x] Naive agent function, different agent types
    + [x] Message Output to independant lists
    + [x] Message Input from independant lists
    + [x] Message Input from the same message list
    + [x] Optional Message Output
    + [x] Agent Birth (independant agent types)
    + [x] Agent Death
    + [x] Device Exception check (only runs with `SEATBELTS` on)
+ Add RTC tests which detect concurrency, through the average runtime of steps:
    + [x] Naive agent function, different agent types
    + Others (see issue)
+ [x] Check synchronisation is still performed as appropriate
+ [x] Disable performance related tests in debug mode.

This has primarily been tested / implemented under Linux. 
+ [x] Validate on Linux 
+ [x] Validate on Windows (WDDM), may be discrepancy.

Ideally the problem sizes should be small enough that individual agent functions will not fill the device, but this requires blocksize information to accurately target. Rather than exposing this, just use very small populations.




### Simple test case
4 agent types, 1 function each in the same layer, with small populations (4k?) each doing lots of pointless global memory accesses and floating point ops to ensure larger run time so that concurrency is (almost) guaranteed to occur. 
Time each function N times, and compare average with and without streams enabled.  

nvprof output for the simple test case, with `NO_SEATBELTS=OFF` showing that concurrency is already being achieved. My test correctly detects this! ~ 70ms with streams, ~ 200ms without
![image](https://user-images.githubusercontent.com/628937/92952663-43cc8e00-f458-11ea-9d92-26af4d4fb784.png)
